File: FRONTEND/src/types/inspection.ts
/**
 * LM-Verify: Legal Metrology (Packaged Commodities) Rules, 2011
 * TypeScript Type Definitions
 */

export type InspectionStatus = 'queued' | 'processing' | 'completed' | 'failed';

export type ComplianceStatus = 'COMPLIANT' | 'NON_COMPLIANT' | 'NEEDS_REVIEW';

export type FindingStatus = 'PASS' | 'FAIL' | 'WARNING' | 'MANUAL_CHECK';

export type FindingSeverity = 'INFO' | 'WARNING' | 'ERROR';

export type UserRole = 'admin' | 'inspector' | 'public_viewer';

export interface UserProfile {
  user_id: string;
  name: string;
  role: UserRole;
  department?: string;
  capabilities: string[];
}

export interface ReadabilityAnalysis {
  readability_score: number;
  estimated_font_size_pt: number;
  min_height_compliant: boolean;
  contrast_score: number;
  rule_9_schedule_ii_status: string;
  notes: string;
}

export interface ProductInfo {
  product_name?: string | null;
  brand?: string | null;
  manufacturer?: string | null;
  mrp?: string | number | null;
  net_quantity?: string | number | null;
  unit?: string | null;
  country_of_origin?: string | null;
  consumer_care?: string | null;
  mfg_date?: string | null;
  batch_no?: string | null;
  manufacturing_date?: string | null;
  expiry_date?: string | null;
  best_before?: string | null;
  email?: string | null;
  phone?: string | null;
  address?: string | null;
  [key: string]: any;
}

export interface ComplianceFinding {
  rule_id: string;
  field: string;
  status: FindingStatus | string;
  severity: FindingSeverity | string;
  message: string;
  rule_title?: string;
  reference_rule?: string; // e.g. "Rule 6(1)(e) Legal Metrology (PC) Rules, 2011"
  source?: {
    document: string;
    reference: string;
    source_url?: string;
    version?: string;
  };
  actual_value?: string | number | null;
  expected?: string | null;
}

export interface InspectionSummary {
  total_rules: number;
  passed: number;
  failed: number;
  warnings: number;
  compliance_percentage?: number;
}

export interface InspectionResult {
  overall_status: ComplianceStatus | string;
  product?: ProductInfo;
  readability?: ReadabilityAnalysis;
  ocr_text?: string;
  confidence?: number;
  findings: ComplianceFinding[];
  summary?: InspectionSummary;
}

export interface Inspection {
  id: string;
  status: InspectionStatus | string;
  created_at: string;
  updated_at?: string;
  image_url?: string;
  image_urls?: string[];
  image_preview?: string;
  image_previews?: string[];
  filename?: string;
  file_size?: number;
  result?: InspectionResult | null;
  error_message?: string | null;
  progress?: number; // 0-100 for processing stage
}

export interface InspectionStats {
  total: number;
  compliant: number;
  non_compliant: number;
  needs_review: number;
  processing?: number;
}

export interface InspectionReport {
  inspection_id: string;
  generated_at: string;
  official_disclaimer: string;
  inspection: Inspection;
  pdf_url?: string | null;
  report_available: boolean;
}

File: BACKEND/app/schemas/inspection.py
from datetime import datetime
from enum import Enum
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, ConfigDict, Field


class InspectionStatus(str, Enum):
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class ComplianceOverallStatus(str, Enum):
    COMPLIANT = "COMPLIANT"
    NON_COMPLIANT = "NON_COMPLIANT"
    NEEDS_REVIEW = "NEEDS_REVIEW"


class RuleStatus(str, Enum):
    PASS = "PASS"
    FAIL = "FAIL"
    WARNING = "WARNING"
    NOT_CHECKED = "NOT_CHECKED"


class RuleSeverity(str, Enum):
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"


class CheckType(str, Enum):
    REQUIRED_FIELD = "REQUIRED_FIELD"
    FIELD_PRESENT = "FIELD_PRESENT"
    FIELD_PATTERN = "FIELD_PATTERN"
    FIELD_VALUE = "FIELD_VALUE"
    CUSTOM = "CUSTOM"


class RuleSource(BaseModel):
    document: str
    reference: str
    source_url: Optional[str] = None
    version: Optional[str] = None
    effective_from: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class RuleDefinition(BaseModel):
    rule_id: str
    name: str
    description: str
    field: str
    check_type: CheckType
    severity: RuleSeverity
    source: RuleSource
    enabled: bool = True
    parameters: Dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(from_attributes=True)


class ExtractedProductData(BaseModel):
    product_name: Optional[str] = None
    brand: Optional[str] = None
    manufacturer: Optional[str] = None
    mrp: Optional[str] = None
    net_quantity: Optional[str] = None
    unit: Optional[str] = None
    country_of_origin: Optional[str] = None
    consumer_care: Optional[str] = None
    mfg_date: Optional[str] = None
    batch_number: Optional[str] = None
    best_before: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ReadabilityAnalysis(BaseModel):
    readability_score: float = Field(ge=0.0, le=100.0, default=95.0)
    estimated_font_size_pt: float = Field(default=8.5)
    min_height_compliant: bool = True
    contrast_score: float = Field(ge=0.0, le=100.0, default=90.0)
    rule_9_schedule_ii_status: str = "COMPLIANT"
    notes: str = "Declaration numeral heights and clarity satisfy Rule 9 & Schedule II minimum standards."

    model_config = ConfigDict(from_attributes=True)


class ComplianceFinding(BaseModel):
    rule_id: str
    field: str
    status: RuleStatus
    severity: RuleSeverity
    message: str
    source: Optional[RuleSource] = None

    model_config = ConfigDict(from_attributes=True)


class ComplianceResult(BaseModel):
    overall_status: ComplianceOverallStatus
    confidence: float = Field(ge=0.0, le=1.0)
    product: ExtractedProductData
    readability: Optional[ReadabilityAnalysis] = None
    ocr_text: Optional[str] = None
    findings: List[ComplianceFinding] = []

    model_config = ConfigDict(from_attributes=True)


class InspectionCreateRequest(BaseModel):
    image_url: Optional[str] = None
    image_urls: List[str] = Field(default_factory=list)


class InspectionCreateResponse(BaseModel):
    id: str
    status: InspectionStatus
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class InspectionDetailResponse(BaseModel):
    id: str
    status: InspectionStatus
    created_at: datetime
    updated_at: Optional[datetime] = None
    image_url: Optional[str] = None
    image_urls: List[str] = Field(default_factory=list)
    result: Optional[ComplianceResult] = None
    error_message: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class HealthResponse(BaseModel):
    status: str
    version: Optional[str] = None
    timestamp: Optional[datetime] = None
